### PR TITLE
Move DataFrame to machinelearning

### PR DESCRIPTION
**Pull request by pgovind**
Move MDA, MDAI, MDA.Tests and MDAI.Tests to machinelearning. 

For an easier time reviewing, the real changes are after the `Merge branch 'port' of ../corefxlab into DataFrame_1` commit.

Just for posterity, I used `git-filter-repo` to pare down corefxlab locally to just the MDA and MDAI parts that I was interested in and then moved them over to the machinelearning repo. `git-filter-repo` was way faster than `git filter-branch` on my machine.

### CODE REVIEW

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:5]: Any reason to pin this here.  Can we let it float to arcade default?

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:8]: We should try to decide how we want to describe package version over in the new repo.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:9]: This is rather odd.  I'd expect us to just use the default.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:27]: This is pretty hacky.  @Anipik has been looking at defining patterns for dotnet/runtime pkgproj -> csproj transition.  I wonder if he has an idea for a better pattern for this case.

I see you're taking the one configuration of MS.D.A.Interactive, putting that assembly in `interactive-extensions/dotnet` and ignoring any of its dependencies.  Do you have any more info to share about how the `interactive-extensions/dotnet` folder works?  What is the expectation of components that use that convention?

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:10]: No hardcoded versions in project files.  We should be defining centrally.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:12]: This is odd, you can use `ExcludeAssets="compile"` instead if that's what you were trying to do.  In general we prefer attributes over child elements for metadata.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:51]: More hardcoded versions here that need to be removed.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:56]: I think this one can now be a ProjectReference.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (michaelgsharp) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:10]: All the version info is in this file. https://github.com/dotnet/machinelearning/blob/master/eng/Versions.props

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:27]: There isn't great documentation on this yet.

You can see some docs at https://github.com/dotnet/interactive/tree/main/docs#extending-net-interactive. Note there is an unlinked "Publishing your extension using NuGet" article that looks like it hasn't been written yet.

I think the best "doc" is the sample at:

https://github.com/dotnet/interactive/blob/main/samples/extensions/Library.nuget/Library.nuget.csproj

cc @jonsequitur @colombod @brettfo @LadyNaggaga

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (pgovind) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:5]: Ah we don't need this anymore. Yup, letting arcade control it now

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:8]: You can remove `VersionPrefix` here all together. And Microsoft.Data.Analaysis will just become an "unstable" project in machinelearning. Which means the next time we release it, it will be versioned `0.17.x`.

See

https://github.com/dotnet/machinelearning/blob/8dd47c13bf815d748a4fa78e9313a83827bb2057/eng/BranchInfo.props#L3-L23

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (pgovind) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:12]: Hmmm this was added in https://github.com/dotnet/corefxlab/pull/2931. I'm wondering if it was copied over from a sample in https://github.com/dotnet/interactive/tree/master/samples/extensions but I can't find it anymore. I'll test locally if we even need this anymore

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (pgovind) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:8]: Hmmm, so 1.5.5 exists in both eng/Versions.props and eng/BranchInfo.props. Anyway, it's not important for our purposes here.

For now, using the version in BranchInfo.props is ok (we'd jump from 0.4.0 -> 0.17.6). But, if we wanted to release a new version of DataFrame, I don't think we'd want to increase the version for all the other projects right?

**Comment by pgovind** [comment]
Alright, I addressed this round of feedback. The only open one left is the `Version` for DataFrame.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (ericstj) [inline src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: Tabs hurt my eyes.

Comment (pgovind) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: ugh, this setting must be mis-matched locally on my machine between VSCode and VS :\ Will fix in a bit

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:23]: What's the minimum TargetFramework required for dotnet interactive extensions?  Are dependencies allowed?  Should the extensions deps file be included? 

I think the way this is included now would add the file multiple times if M.D.A had multiple TargetFrameworks.  I think there might be a different hook other than `TfmSpecificPackageFile` to use for this scenario.  I'll see if I can come up with a sample.

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:23]: > What's the minimum TargetFramework required for dotnet interactive extensions? 

You need to reference: https://www.nuget.org/packages/Microsoft.DotNet.Interactive/1.0.0-beta.21155.3, which is a `netstandard2.1` library. So I guess that's the minimum? dotnet/interactive has already gone to `net5.0`, so I think we could bump this from `netcoreapp3.1` to `net5.0`.

> Are dependencies allowed?

I assume any dependencies you have outside of your NuGet package that aren't already dependencies of dotnet/interactive need to be included. But as it is now, `Microsoft.Data.Analysis.Interactive` doesn't have any external dependencies.

Comment (ericstj) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:23]: I poked around the source and it looks like it just does a LoadFrom: https://github.com/dotnet/interactive/blob/6fd92eb9d72c119b6e7822ca87d96427399a3307/src/Microsoft.DotNet.Interactive/Extensions/AssemblyBasedExtensionLoader.cs#L85

I wonder how that plugin then finds the right copy of M.D.A 🤔   Anyhow, that's not what this comment was about.  That's a thought for another day.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (eerhardt) [inline src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:6]: SA1028 is about code comments not formatting correctly. That seems like something we should be able to fix (either here or in a follow up PR)

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:7]: > But, if we wanted to release a new version of DataFrame, I don't think we'd want to increase the version for all the other projects right?

My thoughts were that we would no longer "release a new version of DataFrame" by itself. It would now start shipping along with the rest of the ML.NET libraries on the same cycle. So when `0.17.7` ships next, we would ship `Microsoft.Data.Analysis` v`0.17.7`, same for `0.17.8`, etc.

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:7]: > so 1.5.5 exists in both eng/Versions.props and eng/BranchInfo.props

It looks like arcade will overwrite `VersionPrefix` if `MajorVersion` and `MinorVersion` are set:

https://github.com/dotnet/arcade/blob/ca7fab569267ed3bc73360882d652d119aae5653/src/Microsoft.DotNet.Arcade.Sdk/tools/Version.BeforeCommonTargets.targets#L76

@michaelgsharp - I think we can delete the `VersionPrefix` line in eng/Versions.props.

Comment (pgovind) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:7]: > So when 0.17.7 ships next, we would ship Microsoft.Data.Analysis v0.17.7, same for 0.17.8, etc.

Perfect. I was thinking we should do a quick MDA 0.5.0 release this month and then get on to the ML releases, but it looks like we can release preview packages whenever we want in ML, so this is not a problem.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (pgovind) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: Out of curiosity, how did you see tabs in this file in GH? I changed all tabs to spaces (no idea how tabs occurred in the first place) in 3 csprojs in the PR, but unfortunately there is no "Change all tabs to spaces in all files" in VS I think. For now, a solution wide regex search yields no tabs in source code

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (pgovind) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:6]: Yup. My plan is to do it in a follow up PR. SA1028 and a bunch of the other warnings are easily fixable

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (michaelgsharp) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj]: @eerhardt I'll run a test build and see. It would be much better to only have to update in 1 place. My local build is having issues though, so I'll test it when I am able to.

**Comment by pgovind** [comment]
Let me know if someone would like something else addressed in this PR. Just waiting for a sign off at this point. I think this is good to go in now.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:54]: Just putting this out there for thought:

We didn't depend on the latest `Microsoft.ML.DataView` before. We depended on `1.0.0`. I wonder if it would be better if we didn't force this dependency to the latest version?

**Review by eerhardt — APPROVED** [review]
Just one comment on a dependency version. Other than that, I think we need to wait for green CI. But everything else looks good.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:54]: Is there any unification or breaking change issues?  In general we've taken an aggressive update cadence with all packages in dotnet.  It helps to ensure folks are patched by default, since Nuget doesn't automatically lift transitive dependencies.

I think someone can still use an old Microsoft.ML with a new DataView right?

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (ericstj) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj]: I was hacking around a bit with this and found that you could do this like the following which is marginally better:
```xml
    <BeforePack>$(BeforePack);IncludeInteractiveExtension</BeforePack>
  </PropertyGroup>

  <Target Name="IncludeInteractiveExtension">
    <MSBuild Projects="../ext/ext.csproj"
         Targets="GetTargetPath">
      <Output TaskParameter="TargetOutputs" ItemName="_interactiveExtension" />
    </MSBuild>
    <ItemGroup>
      <Content Include="@(_interactiveExtension)" Pack="True" PackagePath="interactive-extensions/dotnet" />
    </ItemGroup>
  </Target>
```

This works so long as the interactive extension doesn't cross-target, but it seems like that's a requirement already.  If you ever wanted to cross-target you could define a target in the extension project that ran in outer build and invoked the appropriate inner build framework to select the one that goes in the package (and avoid hardcoding the TFM in a different project).

Finally: what actually builds this project during pack?  I see that M.D.A can't take a project reference since M.D.A.I itself references M.D.A.  If you wanted the build to happen during pack you could change the target above to `Build` instead of `GetTargetPath`.

**Comment by eerhardt** [comment]
> Wait I don't expect CI to be green here

We need to wait for the repo itself to be working before merging any more changes into it. If you merged this, there is a large potential that your change will break the branch even more.

File: src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/Microsoft.Data.Analysis.csproj:54]: In general, ML.NET makes heavy usage of `InternalsVisibleTo`.

However, I looked and I don't think that is true for `Microsoft.ML.DataView` because at one point we wanted to split it completely apart from `Microsoft.ML` (we even named it `Microsoft.Data.DataView` for a time). So I _think_ we should be OK with using a new `Microsoft.Data.DataView` with an older `Microsoft.ML`, but I don't think anyone tests it.

File: src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj
Comment (ericstj) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: It looked too indented then I saw it while trying to select white space.

Comment (ericstj) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: Has codeflow been fixed to work better with Github comment threads?  That was a deal breaker in the past.

Comment (eerhardt) [reply src/Microsoft.Data.Analysis.Interactive/Microsoft.Data.Analysis.Interactive.csproj:2]: Yeah, it still has some issues with comment threads. Typically I only create new conversations from CodeFlow. And respond to threads in GH. I usually have them both open on 2 different windows. I just like seeing the full file, searching across the change, seeing the tree view of files, etc in CodeFlow. It's a much better experience for me.

**Comment by pgovind** [comment]
Alright, CI is green! Thanks for the reviews everyone and @michaelgsharp for fixing CI

**Pull request merged**
